# (노트) 참조와 복사, 얕은복사와 깊은복사 

- toc:true
- branch: master
- badges: true
- comments: true
- author: 신록예찬
- hide: false
- categories: [파이썬]

### 참조와 에일리어싱

`-` 아래의 코드를 관찰하자. 

In [41]:
a=[1,2,3] 
b=a 
a=a+[4] 

현재 a의 출력결과는? 

In [44]:
a

[1, 2, 3, 4]

현재 b의 출력결과는? 

In [45]:
b

[1, 2, 3]

당연하다. 

`-` 이제 다시 아래의 코드를 관찰하자. 

In [46]:
a=[1,2,3]
b=a
a.append(4)

현재 a의 출력결과는? 

In [47]:
a

[1, 2, 3, 4]

현재 b의 출력결과는? 

In [48]:
b

[1, 2, 3, 4]

이거 왜이러지?? 
```python
b=[1,2,3] 
```
아니야? 

`-` 다시 아래의 코드를 살펴보자. 

In [55]:
a=[1,2,3]
b=a
a.append(4)

a,b라는 변수들은 메모리에 어떻게 저장되어있을까? 

상상력을 조금 발휘하면 아래와 같이 여길 수 있다. 

(1) 메모리는 방이 여러개 있는호텔이라고 생각하자. (방은 변수를 저장할 수 있는 공간이다)

(2) 아래를 실행했을 때 
```python 
a=[1,2,3]
```
- 메모리주소1에 존재하는 방을 앞으로 a라고 부르자. 그리고 그 방에 [1,2,3]을 넣는다. 

(3) 아래를 실행했을 때 
```python
b=a
```
- 메모리주소2에 존재하는 방을 b라고 부르고 그 방에 a를 넣는다. 
- 그런데 a=[1,2,3]이므로 b라는 방 역시 [1,2,3]이 있다. 

(4) 아래를 실행했을 때 
```python
a.append(4)
```
- 방 a로 가서 [1,2,3]을 [1,2,3,4]로 바꾼다. 
- 방 b는 아무일도 일어나지 않는다. 



`-` R에서는 맞는 비유인데 파이썬에서는 비유이다. 왜? 

In [62]:
id(a) ## 변수 a가 저장된 메모리주소(=방번호)를 출력하는 함수 

140628239102272

In [63]:
id(b) ## 변수 b가 저장된 메모리주소(=방번호)를 출력하는 함수 

140628239102272

실제로는 변수 a,b의 방주소가 동일하기 때문이다. 

`-` 따라서 파이썬에서는 아래와 같은 비유가 더 적절하다. 

(1) 메모리는 방이 여러개 있는호텔이라고 생각하자. (방은 변수를 저장할 수 있는 공간이다)

(2) 아래를 실행했을 때 
```python 
a=[1,2,3]
```
- 메모리주소1(=140628239102272)에서 [1,2,3]을 만든다. 
- 메모리주소1(=140628239102272)의 방문에 a라는 포스트잇을 붙인다. 
- 따라서 앞으로 [1,2,3]을 찾으려면 a라고 적힌 포스트잇을 찾으면 된다.

(3) 아래를 실행했을 때 
```python
b=a
```
- b라는 포스트잇을 새로 만들고 그 포스트잇은 a와 같은 효과를 주도록 한다. 
- 쉽게말하면 b라는 포스트잇을 만들고 메모리주소1(=140628239102272)의 방문에 b라는 포스트잇을 붙인다는 의미이다. 
- 따라서 [1,2,3]을 찾으려면 a라는 포스트잇을 찾거나 b라는 포스트잇을 찾으면 된다. 

(4) 아래를 실행했을 때 
```python
a.append(4)
```
- a라는 포스트잇이 붙어있는 방으로 찾아가서 [1,2,3]을 찾고 거기에서 append함수를 써서 [1,2,3,4]로 바꾼다. 
- 같은방에 a,b라는 포스트잇이 붙어이므로 b라는 포스트잇을 찾아가서 내용을 열어보면 [1,2,3,4]가 나온다. 

#### 할당문(=)의 이해

`-` 파이썬에서 할당문을 이해하려면 언제나 오른쪽을 먼저 읽어야 한다. 
- 할당문의 오른쪽에서 객체를 생성하거나 가져온다. 
- 그 후에 레이블을 붙이듯이 할당문 왼쪽의 변수가 객체에 바인딩 된다. (이걸 참조라고 부름) 

`-` `b=a`는 

> 나는 이미 a가 의미하는걸 알고 있어, 그런데 a가 의미하는걸 b라고도 부르고싶어

라는 것과 같다. 즉 이미 a라고 부르고 있는것을 가져와서 b라고도 부르고 싶다는 의미인데, 이러한 관점에서 b는 a의 별칭(alias)이라고 볼 수 있다. 

`-` 반대로 생각하면 a역시 b의 별명이라고 볼 수 있다. 

`-` 하나의 메모리 주소에 변수의 이름을 바인드 하는 것을 참조라고 하고, 하나의 메모리 주소에 여러개의 변수이름을 바인드 하는 것을 aliasing이라고 한다. 

#### id, value 

`-` 아래의 예제를 살펴보자. 

In [64]:
a=[1,2,3]
b=a
a.append(4)
c=[1,2,3,4] 

여기에서 a,b,c 는 모두 같은 value를 가진다. 

In [65]:
a

[1, 2, 3, 4]

In [66]:
b

[1, 2, 3, 4]

In [67]:
c

[1, 2, 3, 4]

하지만 그 id까지 같은 것은 아니다. 

In [68]:
id(a)

140628238319424

In [69]:
id(b)

140628238319424

In [70]:
id(c)

140628238319360

`-` 이제 다시 아래의 코드를 살펴보자. 

In [94]:
a=[1,2,3] ### 어떤공간에 [1,2,3]이 저장됨. 공간에 a라는 포스트잇을 붙임 
b=a  ### 그 공간에 b라는 포스트잇도 같이 붙임 
a.append(4) ### a라는 포스트잇이 붙은 공간에 가서 append함수를 적용 

In [95]:
a

[1, 2, 3, 4]

In [96]:
b

[1, 2, 3, 4]

(해설)

In [97]:
a=[1,2,3]
id(a)

140628238373568

In [98]:
b=a
id(b)

140628238373568

`-` 그리고 아래의 코드도 살펴보자. 

In [82]:
a=[1,2,3] ### 어떤공간에 [1,2,3]이 저장됨. 공간에 a라는 포스트잇을 붙임 
b=a ### 그 공간에 b라는 포스트잇도 같이 붙임   
a=[1,2,3]+[4] ### 어떤공간에 [1,2,3]+[4]의 계산결과가 저장됨, 그 공간에 a라는 포스트잇이 붙음 (이 과정에서 첫단계에서 생성된 a라는 포스트잇은 떼어짐)

In [83]:
a

[1, 2, 3, 4]

In [84]:
b

[1, 2, 3]

(해설)

In [85]:
a=[1,2,3]
id(a)

140628238316288

- 공간 `140628238316288`에 [1,2,3]이 생성됨. 공간 `140628238316288`에 a라는 포스트잇붙임

In [86]:
b=a
id(b)

140628238316288

- 공간 `140628238316288`에 b라는 포스트잇도 붙임 

In [90]:
a=[1,2,3]+[4]
id(a)

140628238804032

- [1,2,3]+[4]의 계산결과가 새로운 공간 `140628238804032`에 저장됨. 그 공간에 a라는 포스트잇을 붙임. 
- 따라서 공간 `140628238316288` 에 붙어있던 포스트잇 a를 떼어서 공간 `140628238804032`에 붙임 

In [91]:
id(b)

140628238316288

- 공간 `140628238316288`에는 여전히 b라는 포스트잇이 남아있음 

### 이터닝 

`-` 참조의 개념은 확실하게 알았음. 이제 아래의 예제를 살펴보자. 

In [38]:
a=1+2
id(a)

140628293658992

In [39]:
b=4-1
id(b)

140628293658992

이거 왜 똑같지? 

(해설) 

효율성 때문에 파이썬에서는 내부에 [-5,256]까지의 정수를 미리 저장해둠. 

In [105]:
id(22) ## 이거 5번실행 

140628293659600

`-` [-5,256]이외에는 상식과 동일하게 동작함.

In [143]:
a=1+2222 
id(a)

140628104598768

In [144]:
b=2224-1 
id(b)

140628104598640

In [149]:
id(2223) ## 이거 5번실행

140628104599056

### copy

`-` 참조를 제대로 이해했다면 아래의 예제는 자연스러운 결과임을 알 수 있다. 

In [236]:
l1 = [3,[66,55,44]]
l2 = l1 ## 참조 

In [237]:
l1[0]=4 

In [238]:
l1

[4, [66, 55, 44]]

In [239]:
l2

[4, [66, 55, 44]]

In [240]:
l2.append(5)

In [241]:
l1

[4, [66, 55, 44], 5]

In [242]:
l2

[4, [66, 55, 44], 5]

`-` R과 같은 방식으로 =를 사용하고 싶다면? 

In [243]:
l1 = [3,[66,55,44]]
l2 = l1.copy() ## 카피

In [244]:
id(l1),id(l2) ## 주소가 다르다!!

(140628104170240, 140628104171072)

In [245]:
l1[0]=4

In [247]:
l1

[4, [66, 55, 44]]

In [249]:
l2

[3, [66, 55, 44]]

`-` 이제 다 이해했다고 생각했는데.. 

In [250]:
l1 = [3,[66,55,44]]
l2 = l1.copy()

In [251]:
l1[1].append(33)

In [253]:
l1

[3, [66, 55, 44, 33]]

In [254]:
l2

[3, [66, 55, 44, 33]]

왜 참조한것 마냥 l1와 l2가 같이 바뀌는거야? 

`-` 위의 문제를 해설하여 보자. (주의: 해설이 좀 길어요) 

(상황1) 우선 아래의 상황을 다시 관찰하자. 

In [255]:
a=2222
b=2222

In [256]:
id(a)

140628104166128

In [257]:
id(b)

140628104165840

현재 메모리상황은 아래와 같다. 
- -5~255 까지의 숫자가 저장되어 있음 
- 2222라는 오브젝트가 어떤 공간에 생성되었고 a라는 라벨이 붙음 
- 2222라는 오브젝트가 또다른 어떤 공간에 생성되었고 b라는 라벨이 붙음 

즉 -5~255 이외의 2개의 메모리 공간을 추가사용

(상황2) 이제 아래의 상황을 살펴보자. 

In [279]:
a=[1,2,2222]
b=[1,2,2222]

In [280]:
id(a)

140628104170304

In [281]:
id(b)

140628104610048

In [282]:
id(a[0])

140628293658928

In [283]:
id(b[0])

140628293658928

In [284]:
id(a[1])

140628293658960

In [285]:
id(a[1])

140628293658960

In [286]:
id(a[2])

140628104217584

In [287]:
id(b[2])

140628104218032

In [288]:
a.append(4)

In [289]:
a

[1, 2, 2222, 4]

In [290]:
b

[1, 2, 2222]

해설은 아래와 같다 

- -5~255까지의 숫자가 메모리에 저장되어 있음 
- 2222라는 오브젝트가 어떤 공간에 만들어짐. 공간 X1이라 하자. 
- 리스트 오브젝트가 만들어지고 원소로 1,2,2222를 가짐. 이 공간을 a로 명명함. 
- 2222라는 오브젝트가 또다른 공간에 만들어짐. 공간 X2라고 하자. 
- 리스트 오브젝스가 만들어지고 원소로 1,2,2222를 가짐. 이 공간을 b로 명명함. 
- 공간 a에 원소 4를 추가시킴

즉 -5~255외에 4개의 메모리공간을 추가사용 

- X1: 2222라는 값이 저장됨 
- X2: 2222라는 값이 저장됨
- a: 1,2,공간X1의값
- b: 1,2,공간X2의값

(상황3) 이제 아래의 상황을 살펴보자. 

In [295]:
a=[1,2222,[1,2,3]]
b=a.copy()

In [296]:
id(a)

140628104204608

In [297]:
id(b)

140628104192192

In [294]:
id(a[0])

140628293658928

In [298]:
id(b[0])

140628293658928

In [299]:
id(a[1])

140628104218896

In [300]:
id(b[1])

140628104218896

In [301]:
id(a[2])

140628104188288

In [302]:
id(b[2])

140628104188288

해설을 해보면 아래와 같다. 

- -5~255까지의 숫자가 메모리에 채워져 있음 
- 2222라는 오브젝트가 생성됨. 메모리 공간을 추가로 차지함. 이 공간을 가칭 X1 이라고 하자. 
- 리스트를 담아야할 어떠한 공간 X2가 생성됨. 여기에 원소가 1,2,3으로 추가됨. 
- 리스트를 담아야할 어떠한 공간이 형성되고 여기에서 1,2222,[1,2,3]이 원소로 들어옴. 이 공간을 a라고 하자. 
- b가 정의될 어떠한 공간이 형성됨. b의 원소로는 아까 X1,X2에서 정의된 2222, [1,2,3]과 같은 값들이 재활용되어 원소로 들어옴. 

결국 -5~255외에 추가적인 공간은 아래와 같다. 

- 공간X1: 2222가 저장됨 
- 공간X2: [1,2,3] 이 저장되어 있음 
- 공간a: 1, 공간X1의값, 공간X2의값이 저장됨. 
- 공간b: 1, 공간X1의값, 공간X2의값이 저장됨. 

이상태에서 아래와 같이 한다면? 

In [305]:
a.append(4)

In [306]:
a

[1, 2222, [1, 2, 3], 4]

In [307]:
b

[1, 2222, [1, 2, 3]]

현재메모리상황

- 공간X1: 2222가 저장됨 
- 공간X2: [1,2,3] 이 저장되어 있음 
- 공간a: [1, 공간X1의값, 공간X2의값, 4]
- 공간b: [1, 공간X1의값, 공간X2의값] 

In [308]:
a[2].append(4)

In [309]:
a

[1, 2222, [1, 2, 3, 4], 4]

In [310]:
b

[1, 2222, [1, 2, 3, 4]]

현재메모리상황

- 공간X1: 2222가 저장됨 
- 공간X2: [1,2,3] $\to$ [1,2,3,4]로 변경됨 
- 공간a: [1, 공간X1의값, 공간X2의값, 4]
- 공간b: [1, 공간X1의값, 공간X2의값] 

`-` 결국 아래 2개의 코드는 좀 다르다. 

```python
### 코드1
a=[2222,3333]
b=[2222,3333]
### 코드2
a=[2222,3333]
b=a.copy()
```

(코드1 확인)

In [322]:
a=[2222,3333]
b=[2222,3333] 

In [323]:
id(a), id(b)

(140628104223360, 140628104232576)

In [324]:
id(a[0]), id(b[0])

(140628104251952, 140628104252112)

In [325]:
id(a[1]), id(b[1])

(140628104251984, 140628104252080)

(코드2 확인)

In [326]:
a=[2222,3333]
b=a.copy()

In [327]:
id(a),id(b)

(140628104190208, 140628104205056)

In [328]:
id(a[0]),id(b[0])

(140628104252592, 140628104252592)

In [329]:
id(a[1]),id(b[1])

(140628104252016, 140628104252016)

`-` 결국 `*.copy()` 메소드는 깊은곳까지 복사하지 않는다. (메모리를 너무 아낀다. 짜증나게..) 

`-` 이러한 copy()를 shallow copy 라고 부른다. 

`-` 사실 shallow copy가 평소에는 티가 잘 안난다. 

In [330]:
a=[2222,3333]
b=a.copy()

In [331]:
b[1]=4444 

In [332]:
a

[2222, 3333]

In [333]:
b

[2222, 4444]

(해설) 

- -5~255에 값이 저장되어 있음 
- 2222, 3333을 저장할 메모리공간 2개를 만듬 
- a를 저장할 메모리공간을 만듬 
- b를 저장할 메모리공간을 만들고 2222,3333의 값은 재활용 
- 4444를 따로 만듬. 이를 b의 첫번째원소에 대입

> note: b[1]와 a[1]의 메모리주소는 다르겠지만 b[0]와 a[0]의 메모리주소는 여전히 같을것임 

In [335]:
id(a[1]),id(b[1])

(140628104251856, 140628104252368)

In [336]:
id(a[0]),id(a[0])

(140628104253072, 140628104253072)

그래서 따지고 보면 아래의 코드보다 더 효율적으로 메모리를 씀 

In [337]:
a=[2222,3333]
b=[2222,3333]
b[1]=4444

### shallow copy 연습문제

`-` 예제1: 아래코드의 결과를 예측하라. 결과가 나오는 이유를 설명하라. 

In [353]:
l1 = [3,[66,55,44]]
l2 = l1.copy()
l1[1].append(33)

`-` 예제2: 아래코드의 결과를 예측하라. 결과가 나오는 이유를 설명하라. 

In [354]:
l1 = [3,[66,55,44]]
l2 = l1.copy()
l1[1] = l1[1]+[33]

`-` 예제3: (예제2에서 연속으로 실행) 아래코드의 결과를 예측하라. 결과가 나오는 이유를 설명하라. 

In [355]:
l1[1].remove(33)

`-` 예제4: (예제3에서 연속으로 실행) 아래코드의 결과를 예측하라. 1과 다른이유는? 

In [356]:
l1

[3, [66, 55, 44]]

In [357]:
l2

[3, [66, 55, 44]]

In [350]:
l1[1].append(33)

`-` 예제5: 아래코드의 결과를 관찰하라. 결과가 예제1과 같은 예제2와 같은가? 

In [360]:
l1 = [3,[66,55,44]]
l2 = l1.copy()
l1[1] += [33] 

> Note: `+=` 연산자는 `.append()`와 동일한 효과이다.  

`-` 예제6: 아래코드의 결과를 관찰하라. 왜 예제5와 다른결과가 나오는가? 

In [363]:
l1 = [3,(66,55,44)]
l2 = l1.copy()
l1[1] += (33,) 

(해설) 

- 메모리공간에는 -5~255가 저장
- 추가로 (66,55,44)이 저장될 공간이 필요해짐. 이 공간은 공간X1라고 하자. 
- 추가로 [3,공간X1의값] 이 저장될 공간이 필요해짐. 이 공간은 l1이라 하자. 
- 추가로 공간 l2를 만들고 [3,공간X1의값]을 저장. 
- 공간X1의 값에 33을 추가하고 싶음. 그런데 공간X1에 저장된 값은 튜플인데 튜플은 불변형임 
    - $\to$ 값을 못바꿈 
    - $\to$ 할수없이 새로운공간 X2를 만들고 거기에 (66,55,44,33)을 넣고 그것을 l2에 맵핑함. 

결국 아래와 같은 상황이 되었음 

- 공간X1: (66,55,44) 
- 공간X2: (66,55,44,33)
- 공간l1: [3,공간X1의값]
- 공간l2: [3,공간X2의값] (기존에는 [3,공간X1의값이었음]) 


#### 기억할 것

`-` 파이썬은 메모리를 아끼기 위해서 shallowcopy 라는 이상한 행동을 한다.

`-` 오묘하게도 [1,2,3,4,5]와 같이 0차원 자료형으로만 구성된 리스트라면 문제가 되지 않는다. (메모리는 아끼면서 문제가 되지 않는다? 천재인데?)

`-` 그런데 [[1,2],[3,4]]와 같이 리스트에 리스트가 들어간 형태라면 문제가 생긴다. (이건 개선이 필요하겠는걸?)
- 개선1: 깊은복사
- 개선2: numpy 

### deep copy 

`-` 얕은복사가 아니라 내부객체까지 모두 새로 만드는 깊은복사를 하기 위해서는 아래와 같이 copy 모듈을 사용한다. 

In [366]:
import copy 

`-` 깊은복사 예제1

In [367]:
l1 = [3,[66,55,44]]

In [368]:
l2 = copy.deepcopy(l1)

In [369]:
l2[1].append(55)

In [370]:
l1,l2

([3, [66, 55, 44]], [3, [66, 55, 44, 55]])

`-` 깊은복사 예제2

In [384]:
l1 = [3,[66,[55,44]]]

In [385]:
l2 = copy.deepcopy(l1)

In [386]:
id(l1),id(l2) ## shallow copy = 1 level deep copy 

(140628103869696, 140628103867200)

In [387]:
id(l1[1]),id(l2[1]) ## deep copy = 2 level deep copy 

(140628104264832, 140628105151872)

In [388]:
id(l1[1][1]),id(l2[1][1]) ## deep copy = 3 level deep copy 

(140628104231936, 140628104258176)

In [389]:
l2[1][1].append(33)

In [390]:
l1,l2

([3, [66, [55, 44]]], [3, [66, [55, 44, 33]]])